File: backend/src/sources/ugg.js
import * as R from 'ramda';
import { default_schema } from '../shared/index.js';

import { cl, trinksCon } from '../helpers/index.js';
class WebLogger {
  error(...a) {
    console.error('[Championify]', ...a);
  }
  info(...a) {
    console.info('[Championify]', ...a);
  }
  warn(...a) {
    console.warn('[Championify]', ...a);
  }
}
import T from '../translate.js';
import progressbar from '../progressbar.js';
import store from '../store.js';

const logger = new WebLogger();

const UGG_BUILD_MODES = [
  'rankings',
  'overview',
  'ad-overview',
  'ap-overview',
  'tank-overview',
  'crit-overview',
  'lethality-overview',
  'onhit-overview',
];
const UGG_VERSION_URL =
  'https://static.bigbrain.gg/assets/lol/riot_patch_update/prod/ugg/ugg-api-versions.json';
const LINKS = [
  {ugg_name: '1', position: 'TOP' },
  {ugg_name: '2', position: 'JUNGLE' },
  {ugg_name: '3', position: 'MID' },
  {ugg_name: '4', position: 'BOT' },
  {ugg_name: '5', position: 'SUPPORT' },
];

export const source_info = {
  name: 'u.gg',
  id: 'ugg',
};

export async function getVersion() {
  // They got the wonderful idea NOT to sort patches, so it's a bit of a mess.
  // Get Riot Version from the store, get the first 2 patches lines (eg. 15.12.1 => 15.12)
  // and replace the dot by an underline, to get the proper key.
  const riot_version = store.get('riot_ver')?.split('.').slice(0, 2).join('_');
  return fetch(UGG_VERSION_URL)
    .then(res => res.json())
    .then(data => {
      const v = data[riot_version]?.builds;
      if (v) {
        store.set('ugg_ver', v);
        return v;
      }
      logger.warn('Could not extract version from UGG response, using fallback');
      store.set('ugg_ver', '1.5.0');
      return '1.5.0';
    })
    .catch(err => {
      logger.warn('Error fetching UGG version, using fallback', err);
      store.set('ugg_ver', '1.5.0');
      return '1.5.0';
    });
}

function extractBuildsFromUGG(riotJson, role) {
  const builds = [];
  const layers = Object.values(riotJson || {});

  for (const layer of layers) {
    if (typeof layer !== 'object') continue;
    for (const sub of Object.values(layer)) {
      if (!Array.isArray(sub)) continue;
      for (const build of sub) {
        if (build?.blocks && (!role || build.position === role)) {
          builds.push({ position: build.position, blocks: build.blocks });
        }
      }
    }
  }

  return builds;
}

function translateBlocks(blocks, itemDB) {
  return blocks.map(block => ({
    type: block.type,
    items: block.items.map(it => ({
      id: it.id.toString(),
      count: it.count,
      name: itemDB.data[it.id.toString()]?.name || 'Unknown',
    })),
  }));
}

export async function getSr(gameType = 'ranked_solo_5x5') {
  return (store.get('ugg_ver') ? Promise.resolve(store.get('ugg_ver')) : getVersion())
    .then(async version => {
      const riot_version = store.get('riot_ver');
      const overviewUrl = `https://static.bigbrain.gg/assets/lol/riot_static/${riot_version}/data/en_US/champion.json`;
      const res = await fetch(overviewUrl, { headers: { 'Content-Type': 'application/json' } });
      const data = await res.json();
      return { version, data };
    })
    .then(({ version, data }) => {
      if (!data || !Object.keys(data.data || {}).length)
        throw new Error('Invalid UGG champions response');

      const champions = Object.keys(data.data)
        .map(key => {
          if (key && typeof key === 'string') {
            const champData = data.data[key];
            return { name: key, id: champData.key };
          }
          return null;
        })
        .filter(Boolean)
        .sort((a, b) => a.name.localeCompare(b.name));

      return Promise.all(
        champions.map(async champObj => {
          const { name: champ, id: champId } = champObj;
          try {
            cl(`${T.t('processing')} UGG: ${champ}`);
          } catch {}

          progressbar.incrChamp();

          const riot_version = store.get('riot_ver')?.split('.').slice(0, 2).join('_');
          const url = `https://stats2.u.gg/lol/1.5/${UGG_BUILD_MODES[1]}/${riot_version}/${gameType}/${champId}/${version}.json`;

          try {
            const res = await fetch(url, { headers: { 'Content-Type': 'application/json' } });
            const riotJson = await res.json();

            if (!riotJson || typeof riotJson !== 'object') {
              logger.warn(`Invalid response for ${champ}`);
              return [];
            }

            // --- MIDDLEWARE TRANSLATOR ---
            const builds = extractBuildsFromUGG(riotJson, LINKS.ugg_name);
            if (!builds.length) {
              logger.warn(`No builds parsed for ${champ}`);
              return [];
            }

            const entries = builds.map(build => {
              // convert U.GG blocks → Riot-readable blocks
              const blocksProcessed = trinksCon(
                Array.isArray(build.blocks) ? build.blocks.slice() : []
              );

              const riot_json = R.mergeRight(R.clone(default_schema, true), {
                champion: champ,
                title: `UGG ${champ} ${version}${
                  build.position && build.position !== 'All' ? ` (${build.position})` : ''
                }`,
                blocks: blocksProcessed,
              });

              const file_prefix = (build.position || 'All').toString().toLowerCase();
              return { champ, file_prefix, riot_json, source: 'ugg' };
            });

            return entries;
          } catch (err) {
            logger.error(err);
            store.push('undefined_builds', { source: source_info.name, champ, position: 'All' });
            return [];
          }
        })
      )
        .then(results => results.flat().filter(Boolean))
        .then(data => store.push('sr_itemsets', data));
    });
}

File: src-tauri/src/scrapers/ugg.rs
use async_trait::async_trait;
use reqwest::Client;
use serde_json::Value;
use serde::Deserialize;
use std::collections::HashMap;

use crate::riot_cdn::get_riot_version;
use crate::scrapers::{BuildScraper, ScraperContext};
use crate::AppError;
use crate::scrapers::types::{SourceInfo, ItemBlock, RiotJsonSR, BuildResult};
use crate::get_riot_champions;
use crate::scrapers::ScraperHealth;

#[derive(Deserialize, Debug)]
struct UggBuildItem {
    position: Option<String>,
    blocks: Vec<ItemBlock>,
}

type UggResponse = HashMap<String, HashMap<String, Vec<UggBuildItem>>>;

const UGG_VERSION_URL: &str = "https://static.bigbrain.gg/assets/lol/riot_patch_update/prod/ugg/ugg-api-versions.json";
const UGG_BUILD_MODES: [&str; 8] = [
    "rankings",
    "overview",
    "ad-overview",
    "ap-overview",
    "tank-overview",
    "crit-overview",
    "lethality-overview",
    "onhit-overview",
];

pub fn source_info() -> SourceInfo {
    SourceInfo {
        name: "u.gg".to_string(),
        id: "ugg".to_string(),
    }
}

pub struct UggScraper;

#[async_trait]
impl BuildScraper for UggScraper {
    fn source_info(&self) -> SourceInfo {
        source_info()
    }

    async fn get_version(&self, client: &Client) -> Result<String, AppError> {
        let riot_version = get_riot_version(client).await?;
        let riot_ver_key = riot_version
            .split('.')
            .take(2)
            .collect::<Vec<_>>()
            .join("_");
    
    let response = client
        .get(UGG_VERSION_URL)
        .send()
        .await?
        .error_for_status()?;
    
    let data: HashMap<String, Value> = response.json().await?;
    
    data.get(&riot_ver_key)
        .and_then(|v| v.get("builds"))
        .and_then(|builds| builds.as_str())
        .map(|version_str| version_str.to_string())
        .ok_or_else(|| {
            AppError::Parse(format!(
                "Could not extract version from UGG response for key: {}", 
                riot_ver_key
            ))
        })
    }

    async fn get_sr(
        &self,
        client: &Client,
        context: &ScraperContext,
        _role: Option<&str>,
    ) -> Result<Vec<BuildResult>, AppError> {
        // Basically, UGG has champions from Riot Games DDragon, so no need to fetch anything, just reuse Riot's CDN.
        let ugg_version = self.get_version(client).await?;
        let resolved_riot_version = get_riot_version(client).await?;
        let riot_champions_json = get_riot_champions(client, &resolved_riot_version).await?;
        let champions = riot_champions_json
            .get("data")
            .and_then(|d| d.as_object())
            .ok_or_else(|| AppError::Parse("Invalid UGG champions response".to_string()))?;
    
        let mut all_results = Vec::new();
        
        for (key, champ_data) in champions {
            let champ_id = champ_data
                .get("key")
                .and_then(|k| k.as_str())
                .unwrap_or_default();
            
            println!("Processing UGG: {}", key);
            
            let riot_ver_key = resolved_riot_version
                .split('.')
                .take(2)
                .collect::<Vec<_>>()
                .join("_");
            
            let url = format!(
                "https://stats2.u.gg/lol/{}/{}/{}/{}/{}/{}.json",
                ugg_version.split('.').take(2).collect::<Vec<_>>().join("_"),
                UGG_BUILD_MODES[1],
                riot_ver_key,
                context.game_type.as_str(),
                champ_id,
                ugg_version
            );
            
            match client.get(&url).send().await {
                Ok(response) => {
                    if let Ok(valid_response) = response.error_for_status() {
                        if let Ok(riot_json) = valid_response.json::<UggResponse>().await {
                            let builds = extract_builds(riot_json, None);
                            
                            for (position, blocks) in builds {
                                let file_prefix = position.to_lowercase();
                                let title = format!(
                                    "UGG {} {}{}",
                                    key,
                                    ugg_version,
                                    if position != "All" {
                                        format!(" ({})", position)
                                    } else {
                                        String::new()
                                    }
                                );
                                
                                all_results.push(BuildResult {
                                    champ: key.clone(),
                                    file_prefix,
                                    riot_json: RiotJsonSR {
                                        champion: key.clone(),
                                        title,
                                        blocks,
                                        map: None,
                                    },
                                    source: "ugg".to_string(),
                                });
                            }
                        }
                    } else {
                        log::warn!("Failed to parse JSON into UggResponse for champ {}", key);
                    }
                }
                Err(e) => log::error!("Error fetching builds for {}: {}", key, e),
            }
        }
        Ok(all_results)
    }
    async fn check_health(&self, client: &Client) -> ScraperHealth {
        let url = "https://lol-api-champion.op.gg/api/meta/versions";
        
        match client.get(url).send().await {
            Ok(resp) if resp.status().is_success() => ScraperHealth::Available,
            _ => ScraperHealth::Down,
        }
    }
}

fn extract_builds(data: UggResponse, role: Option<&str>) -> Vec<(String, Vec<ItemBlock>)> {
    let mut builds = Vec::new();
    
    for (_, layer_obj) in data {
        for (_, arr) in layer_obj {
            for build in arr {
                let position = build.position.unwrap_or_else(|| "All".to_string());
                if role.is_none() || role == Some(position.as_str()) {
                    builds.push((position, build.blocks));
                }
            }
        }
    }
    
    builds
}